### ISSUE TITLE
bug: "No discrepancies found" for simple test-case for `hevm equivalence`

### ISSUE THREAD

**Issue by 0xPhaze** [ethereum/hevm#206]
```
hevm version  
0.50.3
```

I'm trying to run `hevm equivalence` on a simple test case and can't get it to work.
My test-case is made up of the following two contracts:

```solidity
contract ArithmeticA {
    function addInt(int256 a, int256 b) public pure returns (int256 c) {
        c = a + b;
    }
}

contract ArithmeticB {
    function addInt(int256 a, int256 b) public pure returns (int256 c) {
        c = a - b;
    }
}
```

I then run:
```
hevm equivalence \
--code-a "608060405234801561001057600080fd5b506004361061002b5760003560e01c80634dad5d1e14610030575b600080fd5b61004a600480360381019061004591906100b1565b610060565b6040516100579190610100565b60405180910390f35b6000818361006e919061014a565b905092915050565b600080fd5b6000819050919050565b61008e8161007b565b811461009957600080fd5b50565b6000813590506100ab81610085565b92915050565b600080604083850312156100c8576100c7610076565b5b60006100d68582860161009c565b92505060206100e78582860161009c565b9150509250929050565b6100fa8161007b565b82525050565b600060208201905061011560008301846100f1565b92915050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052601160045260246000fd5b60006101558261007b565b91506101608361007b565b9250828201905082811215600083121683821260008412151617156101885761018761011b565b5b9291505056fea2646970667358221220e26873223f07ee5e59cdf42ec2f29ab2ed10aabd99829afb55c198d4bfeb688e64736f6c63430008120033" \
--code-b "608060405234801561001057600080fd5b506004361061002b5760003560e01c80634dad5d1e14610030575b600080fd5b61004a600480360381019061004591906100b1565b610060565b6040516100579190610100565b60405180910390f35b6000818361006e919061014a565b905092915050565b600080fd5b6000819050919050565b61008e8161007b565b811461009957600080fd5b50565b6000813590506100ab81610085565b92915050565b600080604083850312156100c8576100c7610076565b5b60006100d68582860161009c565b92505060206100e78582860161009c565b9150509250929050565b6100fa8161007b565b82525050565b600060208201905061011560008301846100f1565b92915050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052601160045260246000fd5b60006101558261007b565b91506101608361007b565b92508282039050818112600084121682821360008512151617156101875761018661011b565b5b9291505056fea26469706673582212208f2e5b6022adb553117e5f4e548ccf992e0fc5614c97349e0fd517af56b38c7764736f6c63430008120033"
```

which outputs:
```
Found 36 total pairs of endstates
Asking the SMT solver for 19 pairs
Reuse of previous queries was Useful in 0 cases
No discrepancies found
```

I've tried using solvers z3 & cvc5  and tried with and without `--sig "addInt(int256,int256)"`.

As a sanity-check, I wrote a test which deploys the above runtime-code in a contract and fuzzes a call to both contracts for equivalence:
```
Running 1 test for test/Arithmetic.equivalence.sol:TestArithmeticEquivalence
[FAIL. Reason: Assertion failed. Counterexample: calldata=0xfd711b910000000000000000000000000000000000000000000000000000000000000000ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff, args=[0, -1]] testAddInt256Equivalence(int256,int256) (runs: 1, μ: 16060, ~: 16060)
Logs:
  Error: Returned value does not match
  Error: a == b not satisfied [bytes]
    Expected: 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
      Actual: 0x0000000000000000000000000000000000000000000000000000000000000001
```

Am I doing something wrong, or is this a bug?

**Comment by msooseth** [comment]
Hi,

Thank you so much for this issue, it's really well-described and easy to reproduce! It is likely  related to #198, and has likely been fixed. However, we'll see if there is more to it than that. Currently,  768efa3feedf3a239df384ead102fa5a20fbc4cc (the tip of `fix-cli-mistakes` branch, the PR branch of #198)  gives:

```
cabal run exe:hevm --  equivalence --code-a "608060405234801561001057600080fd5b506004361061002b5760003560e01c80634dad5d1e14610030575b600080fd5b61004a600480360381019061004591906100b1565b610060565b6040516100579190610100565b60405180910390f35b6000818361006e919061014a565b905092915050565b600080fd5b6000819050919050565b61008e8161007b565b811461009957600080fd5b50565b6000813590506100ab81610085565b92915050565b600080604083850312156100c8576100c7610076565b5b60006100d68582860161009c565b92505060206100e78582860161009c565b9150509250929050565b6100fa8161007b565b82525050565b600060208201905061011560008301846100f1565b92915050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052601160045260246000fd5b60006101558261007b565b91506101608361007b565b9250828201905082811215600083121683821260008412151617156101885761018761011b565b5b9291505056fea2646970667358221220e26873223f07ee5e59cdf42ec2f29ab2ed10aabd99829afb55c198d4bfeb688e64736f6c63430008120033" --code-b "608060405234801561001057600080fd5b506004361061002b5760003560e01c80634dad5d1e14610030575b600080fd5b61004a600480360381019061004591906100b1565b610060565b6040516100579190610100565b60405180910390f35b6000818361006e919061014a565b905092915050565b600080fd5b6000819050919050565b61008e8161007b565b811461009957600080fd5b50565b6000813590506100ab81610085565b92915050565b600080604083850312156100c8576100c7610076565b5b60006100d68582860161009c565b92505060206100e78582860161009c565b9150509250929050565b6100fa8161007b565b82525050565b600060208201905061011560008301846100f1565b92915050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052601160045260246000fd5b60006101558261007b565b91506101608361007b565b92508282039050818112600084121682821360008512151617156101875761018661011b565b5b9291505056fea26469706673582212208f2e5b6022adb553117e5f4e548ccf992e0fc5614c97349e0fd517af56b38c7764736f6c63430008120033"
Up to date
Found 36 total pairs of endstates
Asking the SMT solver for 19 pairs
Reuse of previous queries was Useful in 0 cases
Not equivalent. Counterexample(s):[Cex (SMTCex {vars = fromList [], buffers = fromList [(AbstractBuf "txdata","M\173]\RS\141@\255\255\254\NUL\NUL\NUL\NUL\NUL\NUL\NUL\DC1\193\158U\249\EOT_\248\128\156\192\152\214D\160\DLE\NUL\ACK\NUL8\133\NUL\NUL\NUL\STX\NUL\NUL\NUL\NUL\NUL\NUL\NUL\DC1X]\188\255\255\191\128")], store = fromList [], blockContext = fromList [], txContext = fromList [(CallValue 0,0x0)]}),Cex (SMTCex {vars = fromList [], buffers = fromList [(AbstractBuf "txdata","M\173]\RS$t\NUL$\128\SOH\DC3\175\216<\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\DC1\166\141\245\218G\a\254_\190~\146\SYN\NUL$\128\v\176\152\v\133\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL ")], store = fromList [], blockContext = fromList [], txContext = fromList [(CallValue 0,0x0)]}),Cex (SMTCex {vars = fromList [], buffers = fromList [(AbstractBuf "txdata","M\173]\RS\210\NUL\NUL\NUL\NULI{\\\SOH\NUL\NUL\NUL\NUL\NUL\NUL\NUL\SI\255\255\255\255\255\255\255\255\252\NUL\NUL\NUL\NUL\NUL\128\209\255\192\NUL@\NUL\NUL\NUL\NUL\128\NUL\NUL \NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\EOT\NUL\NUL\NUL\NUL\NUL\DEL")], store = fromList [], blockContext = fromList [], txContext = fromList [(CallValue 0,0x0)]})]
```

Notice that the counterexample (`Cex`) could likely be printed in a way that's more user-friendly.

We'll get back to you soon. Thanks again for the detailed bug report!

**Comment by 0xPhaze** [comment]
Ah great, I also tried running some simple `hevm symbolic` tests that I also couldn't really make sense of. I'll check out the PR once I have time.

**Comment by d-xo** [comment]
Merged #198 to main, fixing the incorrect output, and opened https://github.com/ethereum/hevm/pull/208, which improves the formatting of discovered counterexamples.

Thanks for the report, very much appreciated. Feel free to open further issues, or reach out in [chat](https://matrix.to/#/%23hevm:matrix.org) if you want to discuss your problems with `hevm symbolic` further.
